pkg/sys/userns/idmap.go

```diff
@@ -0,0 +1,133 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+   This file is copied and customized based on
+   https://github.com/moby/moby/blob/master/pkg/idtools/idtools.go
+*/
+
+package userns
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+var (
+	invalidUser = User{Uid: math.MaxUint32, Gid: math.MaxUint32}
+)
+
+// User is a Uid and Gid pair of a user
+//
+//nolint:revive
+type User struct {
+	Uid uint32
+	Gid uint32
+}
+
+// IDMap contains the mappings of Uids and Gids.
+//
+//nolint:revive
+type IDMap struct {
+	UidMap []specs.LinuxIDMapping `json:"UidMap"`
+	GidMap []specs.LinuxIDMapping `json:"GidMap"`
+}
+
+// RootID returns the ID pair for the root user
+func (i IDMap) RootPair() (User, error) {
```

Comment:
Now this is unused, right?

Reply:
When would this be used? For idmap mounts, we want to either honor the mappings or fail. This is to not honor the mappings and find the one for root. When will we use this?

Reply:
Removed

Reply:
Thanks!